incubator/virtualcluster/doc/customresource-syncer.md

```diff
@@ -0,0 +1,185 @@
+
+CRD synchronization ensures all custom defined resource type is deployed in virtual cluster, and CRD cache is properly initialized.
+
+### CRD Cache Remapping
+
+Multi-tenancy virtual cluster infrastructure takes care of CRD cache and CR informer mapping dynamically. When CR syncer sets up its CR informer in virtual cluster, it is possible that CRD has not been deployed in virtual cluster. In this case, once CRD is synced to virtual cluster,  controller-runtime `NewDynamicRESTMapper` is used to dynamically map newly created CRD cache with corresponding CR informer. 
+
+### CRD example
+
+Following Foo CRD will be used here as an example:
+
+```
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type FooSpec struct {
+        Replicas int `json:"replicas"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Foo struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        Spec FooSpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type FooList struct {
+        metav1.TypeMeta `json:",inline"`
+        metav1.ListMeta `json:"metadata,omitempty"`
+        Items []Foo `json:"items"`
+}
+```
+
+## Custom Resource (CR) Syncer 
+
+Following diagram shows CR syncer components and how they interact with multi-tendency infrastructure utilities. 
+
+CR specific synchronization operations are handled by CR Controller, CR Downward syncer (DWS), CR Upward syncer (UWS) (optional) and CR patroller. CR syncer can also include other k8s resources synchronization functionalities by importing corresponding packages from multi-tenancy/incubator/virtualcluster/pkg/syncer/resources .  The Plugin is used to pull all these packages during compilation time.
+
+The Main runs Syncer Server utility: multi-tenancy/incubator/virtualcluster/cmd/syncer/app/server.go, which initializes all pre-included syncer controllers and bootstraps reconciling.
+
+<img width="1176" alt="CR-Syncer-Diagram" src="https://user-images.githubusercontent.com/50972190/112764849-37a4ba00-8fbf-11eb-962d-f2ebcf66406f.png">
+
+### Multi-tenancy Infrastructure Utility Object Construction
+
+CR controller is responsible for instantiating CR specific Mccontroller, Patroller and Listener from following packages:
+
+```
+ Import (
+ "sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/util/mccontroller"
+ "sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/syncer/patrol"
+ "sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/util/listener")
+```
+
+Foo Mccontroller can be constructed as:
+
+```
+multiClusterFooController, err := mccontroller.NewMCController(&v1alpha1.Foo{}, &v1alpha1.FooList{}, c,
+ mc.WithMaxConcurrentReconciles(constants.DwsControllerWorkerLow), mc.WithOptions(options.MCOptions))
+```
```

Comment:
maxconcurrentreconciles is default to `constants.DwsControllerWorkerLow`  so we can remove this option here :)

Reply:
I see

Reply:
Changed